design/230728_tinkerbell_workflow_execution.md

```diff
@@ -0,0 +1,494 @@
+
+The Cancelling state separates Workflows we intend to cancel from those that have been cancelled. This allows Tink Server to issue a cancellation instruction to Tink Agent, and for Tink Agent to indicate successful cancellation. The explicitness of this sequence is useful because:
+* It provides explicit indication that a Running Workflow has stopped making it clear that Tink Agent is available to run a subsequent Workflow.
+* It makes clear to operators the state of the Workflow and why Tink Agent isn't executing a subsequent Workflow.
+
+### Executing Workflows
+
+Tink Server will continue to expose a gRPC API. The protobuf will be simplified to remove superfluous APIs and data structures. Tink Agent will connect to Tink Server to create a long lived command stream. When Tink Server identifies a runnable Workflow for a connected Tink Agent it will dispatch the Workflow on the command stream. Workflows are dispatched in the order they are created. As Tink Agent executes Workflow Actions it will publish events to Tink Server.
+
+<p align="center">
+  <img src="images/230728_tinkerbell_workflow_execution/executing-a-workflow.png" />
+</p>
+
+#### Service protobuf
+
+The Workflow Service exposes 2 APIs, 1 for establishing a command stream and 1 for publishing Workflow execution events. The full protobuf can be found in Appendix A.
+
+```protobuf
+// WorkflowService is responsible for retrieving workflows to be executed by the agent and
+// publishing events as a workflow executes.
+service WorkflowService {
+  // GetWorkflows creates a stream that will receive workflows intended for the agent identified
+  // by the GetWorkflowsRequest.agent_id.
+  rpc GetWorkflows(GetWorkflowsRequest) returns (stream GetWorkflowsResponse) {}
+
+  // PublishEvent publishes a workflow event.
+  rpc PublishEvent(PublishEventRequest) returns (PublishEventResponse) {}
+}
+```
+
+#### Workflow Events
+
+| Event               | Description                                                                                     |
+| ------------------- | ----------------------------------------------------------------------------------------------- |
+| `ActionStarted`     | The action has begun executing                                                                  |
+| `ActionSucceeded`   | The action has completed successfully                                                           |
+| `ActionFailed`      | The action failed to complete. Appropriate failure reason and message will accompany the event  |
+| `WorkflowRejected`  | Tink Agent has explicitly rejected the request to run the Workflow |
+
+### Cancelling a Workflow
+
+Users cancel Workflows by deleting them. In response to the delete request, Tink Controller will update the Workflow to a Cancelling state. Tink Server will watch for Workflows in a Cancelling state. When Tink Server receives an update for an executing Workflow indicating its Cancelling, it will dispatch a cancel command to Tink Agent. 
```

Comment:
I'm going to tweak this to have the Tink Server specify 'Cancelling' because there's too much back and forth between the controller and server and it feels unnecessary. The key is to reflect the state of the system for operators, that's within the scope of Tink Server and it will see a Workflow has been deleted.